fish-rust/src/abbrs.rs

```diff
@@ -0,0 +1,479 @@
+                replacement: abbr.replacement.to_ffi(),
+                is_regex: abbr.is_regex(),
+            })
+        }
+
+        result
+    })
+}
+
+fn abbrs_get_set_ffi<'a>() -> Box<GlobalAbbrs<'a>> {
+    let abbrs_g = abbrs.lock().unwrap();
+    Box::new(GlobalAbbrs { g: abbrs_g })
+}
+
+fn abbrs_replacement_from_ffi(
+    range: SourceRange,
+    text: &CxxWString,
+    set_cursor_marker: &CxxWString,
+    has_cursor_marker: bool,
+) -> abbrs_replacement_t {
+    let mut cursor_marker = None;
+    if has_cursor_marker {
+        let _ = cursor_marker.insert(set_cursor_marker.from_ffi());
+    }
+
+    let replacement = Replacement::from(
+        SourceRange {
+            start: range.start,
+            length: range.length,
+        },
+        text.from_ffi(),
+        cursor_marker,
+    );
+
+    abbrs_replacement_t {
+        range,
+        text: replacement.text.to_ffi(),
+        cursor: replacement.cursor.unwrap_or_default(),
+        has_cursor: replacement.cursor.is_some(),
+    }
+}
+
+pub struct GlobalAbbrs<'a> {
+    g: MutexGuard<'a, AbbreviationSet>,
+}
+
+impl<'a> GlobalAbbrs<'a> {
+    fn add(
+        &mut self,
+        name: &CxxWString,
+        key: &CxxWString,
+        replacement: &CxxWString,
+        position: abbrs_position_t,
+        from_universal: bool,
+    ) {
+        self.g.add(Abbreviation::new(
+            name.from_ffi(),
+            key.from_ffi(),
+            replacement.from_ffi(),
+            position.into(),
+            from_universal,
+        ));
+    }
+
+    fn erase(&mut self, name: &CxxWString) {
+        self.g.erase(&name.from_ffi());
+    }
+}
+use crate::ffi_tests::add_test;
+add_test!("rename_abbrs", || {
+    use crate::{
+        abbrs::{Abbreviation, Position},
+        wchar::L,
+    };
+    pub use widestring::{Utf32Str as wstr, Utf32String as WString};
+
+    with_abbrs_mut(|abbrs_g| {
+        let mut add = |name: &wstr, repl: &wstr, position: Position| {
+            abbrs_g.add(Abbreviation {
+                name: name.into(),
+                key: name.into(),
+                regex: None,
+                replacement: repl.into(),
+                replacement_is_function: false,
+                position,
+                set_cursor_marker: None,
+                from_universal: false,
+            })
+        };
+        add(L!("gc"), L!("git checkout"), Position::Command);
+        add(L!("foo"), L!("bar"), Position::Command);
+        add(L!("gx"), L!("git checkout"), Position::Command);
+        add(L!("yin"), L!("yang"), Position::Anywhere);
+
+        assert!(!abbrs_g.has_name(L!("gcc")));
+        assert!(abbrs_g.has_name(L!("gc")));
+
+        abbrs_g.rename(L!("gc"), L!("gcc"));
+        assert!(abbrs_g.has_name(L!("gcc")));
+        assert!(!abbrs_g.has_name(L!("gc")));
+
+        assert!(!abbrs_g.erase(L!("gc")));
+        assert!(abbrs_g.erase(L!("gcc")));
+        assert!(!abbrs_g.erase(L!("gcc")));
+    })
```

Comment:
Moved these tests from CPP to here, because I needed to create ffi funtions to test these from CPP